guardian/guardian.go

```diff
@@ -0,0 +1,220 @@
+func InitInstance(config Config) {
+	initOnce.Do(func() {
+		if !config.Enabled {
+			log.Warn("Guardian is disabled")
+			return
+		}
+
+		var err error
+		instance, err = newGuardian(config)
+		if err != nil {
+			log.Error("Failed to initialize Guardian", "error", err)
+			return
+		}
+	})
+}
+
+// GetInstance returns the singleton Guardian instance, or an error if it is not initialized.
+func GetInstance() (*Guardian, error) {
+	if instance == nil {
+		return nil, errors.New("guardian is not initialized")
+	}
+
+	return instance, nil
+}
+
+// newGuardian creates a new Guardian instance from the provided config.
+func newGuardian(config Config) (*Guardian, error) {
+	// Create the secure data handler
+	handler, err := securedata.NewPGPSecureHandler(
+		securedata.WithPrivateKeyPath(config.PrivateKeyPath, config.Passphrase),
+		securedata.WithPublicKeyPath(config.PublicKeyPath))
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the bloom filter from file
+	filter, err := store.NewBloomFilterStoreFromFile(
+		config.FilterFilePath,
+		&address.EVMAddressHandler{},
+		store.WithSecureDataHandler(handler))
+	if err != nil {
+		return nil, err
+	}
+
+	// Create file notifier for dynamic filter reload
+	notifier, err := reload.NewFileWatcherNotifier(config.FilterFilePath, 10*time.Second)
+	if err != nil {
+		return nil, err
+	}
+
+	// Start reload manager
+	manager := reload.NewReloadManager(filter, notifier)
+	if err := manager.Start(context.Background()); err != nil {
+		return nil, err
+	}
+
+	log.Info("Guardian initialized", "file", config.FilterFilePath)
+	return &Guardian{
+		filter:  filter,
+		manager: manager,
+	}, nil
+}
+
+// CheckTransaction checks if the sender or recipient in the transaction is in the filter file.
+// Returns true if the transaction interacts with any filtered addresses.
+func (p *Guardian) CheckTransaction(signer types.Signer, tx *types.Transaction) bool {
+	// Extract the sender's address
+	from, err := types.Sender(signer, tx)
+	if err != nil {
+		log.Error("Failed to extract 'from' address", "err", err)
+		return false
+	}
+
+	// Check the sender's address
+	if filtered, err := p.checkAddress(tx, from.Hex(), from.Hex()); err != nil || filtered {
+		if err != nil {
+			log.Error("Error checking sender address", "err", err)
+		}
+		return filtered
+	}
+
+	// Check the recipient's address if applicable
+	if to := tx.To(); to != nil {
+		if filtered, err := p.checkAddress(tx, from.Hex(), to.Hex()); err != nil || filtered {
+			if err != nil {
+				log.Error("Error checking recipient address", "err", err)
+			}
+			return filtered
+		}
+	}
+
+	return false
+}
+
+// Stop shuts down Guardian, stops the filter reload manager safely.
+func (p *Guardian) Stop() {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if p.manager != nil {
+		_ = p.manager.Stop()
+	}
+}
+
+// Reset allows you to reset the Guardian instance.
+// This stops the current instance and resets it to nil.
+func Reset() {
```

Comment:
I assume `Reset` should only be called in unit tests, right? 
If so, move this function to test dir.

Reply:
Done